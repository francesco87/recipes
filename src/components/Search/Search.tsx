import { useQuery } from '@tanstack/react-query';
import { getRecipesByIngredient } from '../../api/api.ts';
import { useEffect, useState } from 'react';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import SearchIcon from '@mui/icons-material/Search';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import RecipeBox from '../RecipeBox/RecipeBox.tsx';
import { useSearchParams } from 'react-router';
import Modal from '../Modal/Modal.tsx';
import Typography from '@mui/material/Typography';

const Search = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const [ingredient, setIngredient] = useState(searchParams.get('search'));
  const [touched, setTouched] = useState(false);
  const { data, isFetching, refetch, error } = useQuery({
    queryKey: ['recipes'],
    queryFn: () => getRecipesByIngredient(ingredient),
    enabled: false,
  });
  const [showModal, setShowModal] = useState(false);
  const recipes = data?.meals || [];
  const closeModal = () => setShowModal(false);

  useEffect(() => {
    setShowModal(!!error);
  }, [error]);

  const NoResults = () => {
    if (data?.meals === null) {
      return (
        <Typography sx={{ textAlign: 'center' }} variant="subtitle1" gutterBottom>
          No recipes found
        </Typography>
      );
    }
    return <></>;
  };

  return (
    <Box>
      <form
        onSubmit={(e) => {
          e.preventDefault();
          if (ingredient) {
            setSearchParams({ search: ingredient || '' });
            refetch();
          }
        }}
      >
        <Grid container spacing={1} sx={{ marginBottom: 4 }}>
          <Grid size={{ xs: 12, md: 9 }}>
            <TextField
              placeholder="Type a meal or an ingredient"
              fullWidth
              size="small"
              label="Search recipes"
              variant="outlined"
              onChange={(e) => {
                setTouched(true);
                setIngredient(e.target.value);
              }}
              value={ingredient || ''}
              disabled={isFetching}
              error={touched && !ingredient}
              helperText={touched && !ingredient ? 'Required' : ''}
            />
          </Grid>
          <Grid size={{ xs: 12, md: 3 }}>
            <Button
              fullWidth
              size="large"
              variant="contained"
              loading={isFetching}
              loadingPosition="end"
              disableElevation
              endIcon={<SearchIcon />}
              type="submit"
            >
              Search
            </Button>
          </Grid>
        </Grid>
      </form>
      <Grid container spacing={1}>
        {recipes?.map((recipe) => (
          <Grid size={{ xs: 12, md: 4 }} key={recipe.idMeal}>
            <RecipeBox recipe={recipe} />
          </Grid>
        ))}
      </Grid>
      <NoResults />
      <Modal
        title="Error"
        message={error?.message}
        showModal={showModal}
        handleClose={closeModal}
      />
    </Box>
  );
};

export default Search;
