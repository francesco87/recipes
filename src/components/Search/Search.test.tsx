import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import Search from './Search';
import type { RecipeType } from '../RecipeBox/RecipeBox.tsx'; // ✅ path corretto

vi.mock('../../api/api.ts', () => ({
  getRecipesByIngredient: vi.fn(),
}));

// ✅ Mock React Query
const mockRefetch = vi.fn();
let mockQueryData: { meals: { idMeal: string; strMeal: string }[] } | null = null;
let mockError: { message: string } | null = null;
let mockIsFetching = false;

vi.mock('@tanstack/react-query', () => ({
  useQuery: () => ({
    data: mockQueryData,
    error: mockError,
    isFetching: mockIsFetching,
    refetch: mockRefetch,
  }),
}));

// ✅ Mock TanStack Router search params
const mockSetSearchParams = vi.fn();
vi.mock('react-router', () => ({
  useSearchParams: () => [new URLSearchParams({ search: 'pasta' }), mockSetSearchParams],
}));

// ✅ Mock RecipeBox
vi.mock('../RecipeBox/RecipeBox.tsx', () => ({
  __esModule: true,
  default: ({ recipe }: { recipe: RecipeType }) => (
    <div data-testid="recipe-box">{recipe.strMeal}</div>
  ),
}));

// ✅ Mock Modal to check if it is visible
vi.mock('../Modal/Modal.tsx', () => ({
  __esModule: true,
  default: ({ showModal }: { showModal: boolean }) =>
    showModal ? <div data-testid="modal">Error</div> : null,
}));

beforeEach(() => {
  mockRefetch.mockClear();
  mockSetSearchParams.mockClear();
  mockQueryData = null;
  mockError = null;
  mockIsFetching = false;
});

describe('Search Component', () => {
  it('Search Component Snapshot', () => {
    const { container } = render(<Search />);
    expect(container).toMatchSnapshot();
  });

  it('Shows validation error when input is empty', () => {
    render(<Search />);

    const input = screen.getByLabelText(/Search recipes/i);
    fireEvent.change(input, { target: { value: '' } });

    const button = screen.getByRole('button', { name: /Search/i });
    fireEvent.click(button);

    expect(screen.getByText('Required')).toBeInTheDocument();
  });

  it('Renders recipes when api returns meals', () => {
    mockQueryData = {
      meals: [
        { idMeal: '1', strMeal: 'Pizza' },
        { idMeal: '2', strMeal: 'Pasta' },
      ],
    };

    render(<Search />);
    expect(screen.getAllByTestId('recipe-box')).toHaveLength(2);
  });

  it('Renders modal if api return an errors', () => {
    mockError = new Error('Network error');

    render(<Search />);
    expect(screen.getByTestId('modal')).toBeInTheDocument();
  });
});
