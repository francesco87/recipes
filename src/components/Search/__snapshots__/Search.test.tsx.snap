// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search Component > Search Component Snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <form>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-1 css-19iv2b-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-12 MuiGrid-grid-md-9 css-1q194yh-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-hvjq6j-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="_r_0_"
              id="_r_0_-label"
            >
              Search recipes
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
                id="_r_0_"
                placeholder="Type a meal or an ingredient"
                type="text"
                value="pasta"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ex8a5f-MuiNotchedOutlined-root"
                >
                  <span>
                    Search recipes
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-lvb8h-MuiGrid-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary MuiButton-disableElevation MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary MuiButton-disableElevation MuiButton-fullWidth css-18unm2x-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Search
            <span
              class="MuiButton-loadingWrapper"
              style="display: contents;"
            />
            <span
              class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeLarge css-3155yg-MuiButton-endIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </form>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-1 css-1yvsbrk-MuiGrid-root"
    />
  </div>
</div>
`;
